keps/sig-node/4639-oci-volume-source/README.md

```diff
@@ -0,0 +1,940 @@
+   - An artifact describes any content that is stored and distributed using the OCI image format.
+     It includes not just container images but also other types of content like Helm charts, WASM modules, machine learning models, etc.
+   - Artifacts use the same image manifest and layer structure but may contain different types of data
+     within those layers. The artifact manifest can have media types that differ from those in standard container images.
+   - **Use Case:** Allows the distribution of non-container content using the same infrastructure and tools developed for OCI images.
+
+**OCI Object:**
+   - Umbrella term encompassing both OCI images and OCI artifacts. It represents any object that conforms to the OCI specifications for storage and distribution.
+
+### Risks and Mitigations
+
+- **Security Risks:** Allowing direct mounting of OCI images introduces potential attack vectors. Mitigation includes thorough security reviews and
+  limiting access to trusted registries. Limiting to OCI artifacts (non-runnable content) or read-only mode may lessen the security risk.
+- **Compatibility Risks:** Existing webhooks watching for the images used by the pod using some policies will need to be updated to expect the image to be specified as a `VolumeSource`.
+- **Performance Risks:** Large images or artifacts could impact performance. Mitigation includes optimizations in the implementation and providing
+  guidance on best practices for users.
+
+## Design Details
+
+The new `VolumeSource` will be defined in the Kubernetes API, and the implementation will involve updating components (CRI, Kubelet)
+to support this source type. Key design aspects include:
+
+- API changes to introduce the new `VolumeSource` type.
+- Modifications to the Kubelet to handle mounting OCI images and artifacts.
+- Handling image pull secrets and registry authentication.
+- Reuse existing logic from ConfigMaps for:
+  - Determining the file location on the host filesystem.
+  - Handling updates to the OCI image or artifact, similar to how ConfigMaps update mounted files.
+
+The following code snippet illustrates the proposed API change:
+
+```yaml
+apiVersion: v1
+kind: Pod
+metadata:
+  name: example-pod
+spec:
+  volumes:
+  - name: oci-volume
+    oci:
+      image: "example.com/my-image:latest"
+      pullPolicy: IfNotPresent
+  containers:
+  - name: my-container
+    image: busybox
+    volumeMounts:
+    - mountPath: /data
+      name: oci-volume
+```
+
+### Kubelet and Container Runtime Interface (CRI) support for OCI artifacts
+
+Kubelet and the Container Runtime Interface (CRI) currently handle OCI images. To support OCI artifacts,
+potential enhancements may be required:
+
+**Extended Media Type Handling in the container runtime:**
+   - Update container runtimes to recognize and handle new media types associated with OCI artifacts.
+   - Ensure that pulling and storing these artifacts is as efficient and secure as with OCI images.
```

Comment:
if the volume source is updated to use a new released model artifact, then these images which were mounted with old artifact should unmount the old one first, and re-mount the new artifact files. 

how to deal with the case if the images are accessing these files duing mount/un-mount?

Reply:
Is this discussion referring to https://github.com/kubernetes/enhancements/pull/4642#discussion_r1626363466?

Reply:
With a CSI approach, we could offer in-place updates (where `inotify` and friends would fire). If the container runtime does the image pull, we don't really have that option.

Reply:
> in-place updates
We could during sync pod, but I assume we left that out intentional as discussed earlier.